### DISCUSSION TITLE
activate_distance in ActivateElementsByPath does not seem to work.

### DISCUSSION THREAD

**Discussion by zx1987** [idaholab/moose#17947] [Q&A Meshing] [answered]
All

I am playing around the user object ActivateElementsByPath following the example https://mooseframework.inl.gov/source/userobject/ActivateElementsByPath.html 

I adapt it and make a example to activate elements within a given distance  along a path. It appears to me activate_distance is not working as expected. While I used different value for activate-distance, it is only activating one layer of elements on top and below the path.  I am attaching a copy of the input file and could anybody advise what might be wrong? Thank you.

```
[Problem]
  kernel_coverage_check = false
[]

[Mesh]
  [./gen]
    type = GeneratedMeshGenerator
    dim = 2
    xmin =0
    ymin=0
    xmax =0.01
    ymax =0.004
    nx=100
    ny= 40
    nz=1
  [../]
  [./bottom_domain]
    input = gen
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0 0 0'
    top_right = ' 0.01  0.001 0'
    block_id = 1
  [../]
  [./top_domain]
    input = bottom_domain
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0 0.001 0'
    top_right = '0.01  0.004 00'
    block_id = 2
  [../]
  [./sidesets]
    input = top_domain
    type = SideSetsAroundSubdomainGenerator
    normal = '1 0 0'
    block = 1
    new_boundary = 'moving_interface'
  []
[]

[Variables]
  [./temp]
    block = '1'
  [../]
[]

[Functions]
  [./fy]
    type = ParsedFunction
    value= 'if(t<12.5,0.0015,0.0025)'
  [../]
  [./fx]
    type = ParsedFunction
    value= 'if(t<12.5,0.0008*t, 0.01-0.0008*(t-12.5))'
  [../]
[]

[Preconditioning]
  [./smp]
    type = SMP
    full = true
  [../]
[]


[Executioner]
  type = Transient

  automatic_scaling = true
  solve_type = 'NEWTON'
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'

  line_search = 'none'

  l_max_its = 10
  nl_max_its = 20
  nl_rel_tol = 1e-4

  start_time = 0.0
  end_time = 25.0
  dt = 1e-1
  dtmin = 1e-4
[]

[UserObjects]
  [./activated_elem_uo]
    type = ActivateElementsByPath
    execute_on = timestep_begin
    activate_distance = 0.0005
    function_x= fx
    function_y= fy
    active_subdomain_id = 1
    expand_boundary_name = 'moving_interface'
  [../]
[]
[Outputs]
  exodus = true
[]

```

Best,
Xiang

**Comment by dewenyushu** [comment, marked as answer]
Hi @zx1987 , thanks for finding this issue. 

I tested this out and realized that this is a bug related to the usage of `close_to_point` function. I created a PR (#17954) which should fix this issue.

Also, I'd like you to be aware of the `CoupledVarThresholdElementSubdomainModifier ` user object, which is a more generalized version of adding/deleting/removing the elements. We also have a plan to combine these capabilities
https://mooseframework.inl.gov/source/userobject/CoupledVarThresholdElementSubdomainModifier.html

**Reply by zx1987** [reply]
Hi @dewenyushu , thanks for the bug fix and pointing me to the `CoupledVarThresholdElementSubdomainModifier`. Will check it out.

Best,
Xiang

**Reply by zx1987** [reply]
Hi @dewenyushu 

I just checked the new code, and have some further questions here:

I added the `activate_distance=9` into the example check_element_addition.i input file, as below
`[Problem]
  kernel_coverage_check = false
[]

[Mesh]
  [./gen]
    type = GeneratedMeshGenerator
    dim = 3
    xmin =0
    xmax =10
    ymin =0
    ymax =10
    zmin =0
    zmax =0.5The width of the newly activated elements do not seem to be correct, as I am expecting any elements within a distance of 9 to the 
    nx=100
    ny=100
    nz=1
  [../]
  [./left_domain]
    input = gen
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0 0 0'
    top_right = '5 10 0.5'
    block_id = 1
  [../]
  [./right_domain]
    input = left_domain
    type = SubdomainBoundingBoxGenerator
    bottom_left = '5 0 0'
    top_right = '10 10 0.5'
    block_id = 2
  [../]
  [./sidesets]
    input = right_domain
    type = SideSetsAroundSubdomainGenerator
    normal = '1 0 0'
    block = 1
    new_boundary = 'moving_interface'
  []
[]

[Variables]
  [./temp]
    block = '1'
  [../]
[]

[Functions]
  [./fx]
    type = ParsedFunction
    value= '7.5'
  [../]
  [./fy]
    type = ParsedFunction
    value= '2.5*t'
  [../]
  [./fz]
    type = ParsedFunction
    value= '0.25'
  [../]
[]

[Preconditioning]
  [./smp]
    type = SMP
    full = true
  [../]
[]


[Executioner]
  type = Transient

  automatic_scaling = true

  solve_type = 'NEWTON'

  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'

  line_search = 'none'

  l_max_its = 10
  nl_max_its = 20
  nl_rel_tol = 1e-4

  start_time = 0.0
  end_time = 1.0
  dt = 1e-1
  dtmin = 1e-4
[]

[UserObjects]
  [./activated_elem_uo]
    type = ActivateElementsByPath
    execute_on = timestep_begin
    function_x= fx
    function_y= fy
    function_z= fz
    activate_distance=9
    active_subdomain_id = 1
    expand_boundary_name = 'moving_interface'
  [../]
[]

[Outputs]
  exodus = true
[]`

I got a snapshot rendered with block ID as below at t=0.3s. The width of the newly activated elements do not seem to be correct, as I am expecting any elements within a distance of 9 to the line of x=7.5 will be activated (white color), and also there is one element in the top row of the newly activated element is shown in red. Did me misunderstand anything here?

Thank you.

Best,
xiang

![image](https://user-images.githubusercontent.com/11442693/120045359-0a7b8480-bfcd-11eb-8b59-ca800cb67be2.png)

**Reply by dewenyushu** [reply]
Hi @zx1987

 I think you are understanding the usage of this pretty well. I suppose the issue may because you did not manage to grab the recent fix. Since it is merged this morning, the fix only appears on the `next` branch now. I would suggest trying rebase your branch on `next`, re-compile, and run your problem again

I just tried running with your above input, and I got something that looks okay. Following is a snapshot at t=0.3s:

![Screen Shot 2021-05-28 at 4 49 49 PM](https://user-images.githubusercontent.com/58825741/120048946-d0fb4700-bfd5-11eb-9942-a6045ab28392.png)

**Reply by zx1987** [reply]
Hi @dewenyushu 

Yes your are right. Wen mentioned this to me earlier, but at that time I already messed up with my previous installation and I had a new install, thinking that the update is already in in. After grabbing the update it worked as expected. Thank you.

Best,
Xiang

**Reply by dewenyushu** [reply]
Sounds great! Glad it worked :)
